api/v1beta1/disruption_types.go

```diff
@@ -411,50 +432,57 @@ func (s *DisruptionSpec) GetDisruptionCount() int {
 
 // RemoveDeadTargets removes targets not found in matchingTargets from the targets list
 func (status *DisruptionStatus) RemoveDeadTargets(matchingTargets []string) {
-	var desiredTargets []string
+	desiredTargets := TargetInjections{}
 
-	for index := 0; index < len(status.Targets); index++ {
-		if utils.Contains(matchingTargets, status.Targets[index]) {
-			desiredTargets = append(desiredTargets, status.Targets[index])
+	targetNames := status.TargetInjections.GetTargetNames()
+
+	for _, matchingTarget := range matchingTargets {
+		if utils.Contains(targetNames, matchingTarget) {
+			desiredTargets[matchingTarget] = status.TargetInjections[matchingTarget]
 		}
 	}
 
-	status.Targets = desiredTargets
+	status.TargetInjections = desiredTargets
 }
 
 // AddTargets adds newTargetsCount random targets from the eligibleTargets list to the Target List
 // - eligibleTargets should be previously filtered to not include current targets
-func (status *DisruptionStatus) AddTargets(newTargetsCount int, eligibleTargets []string) {
+func (status *DisruptionStatus) AddTargets(newTargetsCount int, eligibleTargets TargetInjections) {
 	if len(eligibleTargets) == 0 || newTargetsCount <= 0 {
 		return
 	}
 
-	for i := 0; i < newTargetsCount && len(eligibleTargets) > 0; i++ {
-		index := rand.Intn(len(eligibleTargets)) //nolint:gosec
-		status.Targets = append(status.Targets, eligibleTargets[index])
-		eligibleTargets[len(eligibleTargets)-1], eligibleTargets[index] = eligibleTargets[index], eligibleTargets[len(eligibleTargets)-1]
-		eligibleTargets = eligibleTargets[:len(eligibleTargets)-1]
+	targetNames := eligibleTargets.GetTargetNames()
+
+	for i := 0; i < newTargetsCount && len(targetNames) > 0; i++ {
+		index := rand.Intn(len(targetNames)) //nolint:gosec
+		targetName := targetNames[index]
+		targetNames[len(targetNames)-1], targetNames[index] = targetNames[index], targetNames[len(targetNames)-1]
+		targetNames = targetNames[:len(targetNames)-1]
+
+		status.TargetInjections[targetName] = eligibleTargets[targetName]
+		delete(eligibleTargets, targetName)
 	}
 }
 
 // RemoveTargets removes toRemoveTargetsCount random targets from the Target List
 func (status *DisruptionStatus) RemoveTargets(toRemoveTargetsCount int) {
-	for i := 0; i < toRemoveTargetsCount && len(status.Targets) > 0; i++ {
-		index := rand.Intn(len(status.Targets)) //nolint:gosec
-		status.Targets[len(status.Targets)-1], status.Targets[index] = status.Targets[index], status.Targets[len(status.Targets)-1]
-		status.Targets = status.Targets[:len(status.Targets)-1]
+	targetNames := status.TargetInjections.GetTargetNames()
+
+	for i := 0; i < toRemoveTargetsCount && len(status.TargetInjections) > 0; i++ {
```

Comment:
duplicate logic, consider to extract a private method

Reply:
Good point! 👍